### DISCUSSION TITLE
TODO list from 2/11/2022

### DISCUSSION THREAD

**Discussion by jesse-osborne** [mptoolkit/mptoolkit#16] [General]
`mp-ibc-concatenate` file to join two IBC wavefunctions together.
- We need to make sure that the the boundary iMPSs have the same bases at the unit cell boundary, and that enough sites are added in between the two windows so that the effects of windows on the boundaries are negligible.

ARPACK wrapper:
- Merge version on the ibc-tdvp branch onto main: bffccc4a8291114564d63c79c0e771108a9ebabc, 0e715aa44dc9f50b60f323a1a593bacde985b2a8.
- Fix sorting of eigenvalues when not using largest magnitude.
- Add a wrapper for a Hermitian matrix solver.

**Comment by ianmccul** [comment]
The problem is going to be matching the basis at the join.  If we have a very large number of sites between them then this is exact, but presumably we want the distance between the wavepackets to be as small as possible.

If the right hand side of the left IBC is in the same basis at the unit cell boundary (which it should be, but might not be at the moment!) then we can convert that right hand side into left-ortho form with a sequence of SVD's, and the lambda matrix will eventually converge, but even before it has converged completely we could join on the window and right hand side of the right IBC.  The effect of too-few sites in between will be that the right hand IBC needs to be reorthogonalized.

In principle, if the left window is left-ortho and the right window is right-ortho, we could join them together with no background sites in between, but it has the same issue as iDMRG - there will be two Lambda matrices and how to combine them into 1.  iDMRG uses a recurrence relation to compute $\Lambda_R \Lambda_0^{-1} \Lambda_L$, but here we don't have $\Lambda_0$.  Or do we?  Is it just the $\Lambda$ matrix of the background?

**Comment by jesse-osborne** [comment]
At the moment, I am using something like this to generate the right boundaries for an IBC:
https://github.com/ianmccul/mptoolkit-dev/blob/0e715aa44dc9f50b60f323a1a593bacde985b2a8/mp/mp-right-canonicalize.cpp#L89-L99
Just calling `InfiniteWavefunctionRight PsiRight(PsiLeft);` doesn't seem to ensure that the basis is the same at the boundary, but this way seems to work (at least, I haven't had any obvious problems with it).

And even if they don't have the same basis, would it be possible to find the basis transformation matrix by transforming them both left-orthogonal form, and calculate the left eigenvector of the mixed transfer matrix. (If these were exactly the same state with the same bases, this eigenvector would be the identity matrix, so if they were the same state up to some basis transformation, then this eigenvector will presumably be the unitary matrix which transforms between the bases?)

**Reply by ianmccul** [reply]
Yeah I'm wondering if there is a transfer matrix formulation that would avoid the inverse.

**Comment by jesse-osborne** [comment]
My current idea is this:
If we start with two IBCs $\cdots AAABBBCCC\cdots$ and $\cdots CCCDDDEEE\cdots$, then the user can specify how many sites to add to the left and the right windows from the 'middle' boundary $N_L$ and $N_R$ (usually, they would be the same $N_L = N_R \equiv N/2$, but there might be situations where it might be helpful for them different), so that the final wavefunction will be something like
$$\cdots AAABBB\overbrace{CCC}^{N_L}\overbrace{CCC}^{N_R}DDDEEE\cdots.$$
Then you can left-orthogonalise $BBBCCC$ to $LLLLLLD_LV_L^\dagger$ and right-orthogonalise $CCCDDD$ to $U_RD_RRRRRRR$.

Then the question is how to tell whether the orthogonality centres of the two windows $D_LV_L^\dagger$ and $U_RD_R$ are the same. My idea is to insert $V_L^\dagger V_L$ and $U_R^\dagger U_R$ into the windows to get
$$LLLLLLV_L^\dagger V_LD_LV_L^\dagger = LLLLLL'D_L', \qquad L' = LV_L^\dagger, \qquad D_L' = V_LD_LV_L^\dagger,$$
$$U_RD_RU_R^\dagger U_RRRRRRR = D_R' R'RRRRR, \qquad D_R' = U_RD_RU_R^\dagger, \qquad R' = U_RR.$$
$D_L'$ and $D_R'$ will then be square matrices with the same bases as the boundary $\Lambda$ matrix of the state underlying the middle boundary, and you can check whether $\epsilon_{L/R} \equiv \||D_{L/R}' - \Lambda\||_2$ fall below some threshold. The full window can then be set to
$$LLLLLL'\Lambda R'RRRRR$$
and then be left-orthogonalised and saved.

The default behaviour of the program should then be to pick the minimum $N_{L/R}$ such that $\epsilon_{L/R}$ are below the threshold. $N_{L/R}$ can then be specified as optional parameters, and print a warning message or abort if $\epsilon_{L/R}$ are above the threshold.

**Reply by ianmccul** [reply]
This is where the inverse comes in - $\Lambda' = D'_L \Lambda^{-1} D'_R$.  If you do the check that all of the $\Lambda$'s are within epsilon then you can skip the inversion, but the point about the inverse is that it works very well (at least in the iDMRG context, I expect it would also work quite well here) even if $\Lambda$ is quite different.  In principle you could join the two windows together with no padding sites (although in practice you probably want at least a few).

**Reply by jesse-osborne** [reply]
Would it be possible to try something like $\Lambda' = (D_L')^{1/2} (D_R')^{1/2}$? This way you don't need to use the inverse of $\Lambda$, but you do need to use the square roots of $D_{L/R}'$. Since (I think?) taking the square root of singular values below machine epsilon is safer than inverting them.

(And since $D_{L/R}'$ are unitarily similar to the diagonal matrices $D_{L/R}$,  i.e. $D_L' = V_L D_L V_L^\dagger$, you can write the square root of $D_L'$ as $(D_L')^{1/2} = V_L D_L^{1/2} V_L^\dagger$, and similarly for $(D_R')^{1/2}$.)

**Reply by ianmccul** [reply]
See https://arxiv.org/abs/0804.2509 figure 2 and the discussion on page 6.

**Reply by ianmccul** [reply]
inverting singular values isn't that much of a problem - the physically relevant quantity is the density matrix, which is the square of that.  So a cutoff singular value of $10^{-8}$ or similar is fine.

**Reply by ianmccul** [reply]
But if there is some transfer matrix way of getting the same result, that would be ideal.

**Comment by jesse-osborne** [comment]
I found a better alternative to using ARPACK to minimise the cost function for localising the wavepacket by using `LinearAlgebra::DiagonalizeHermitian` instead b0e5734179d81adcc6e6c6d1d844ce4cc5e7df10. This function calls `zheev` from LAPACK to diagonalise the full matrix, which is much quicker in most cases, and doesn't get stuck, like ARPACK did sometimes.
Since the matrices are dense and have a dimension equal to the number of momentum steps, calculating the full diagonalisation isn't too costly, and so the task 'Add a wrapper for a Hermitian matrix solver' is probably not needed at the moment.

**Reply by ianmccul** [reply]
Yes I was going to suggest that.  But I'll add a hermitian matrix solver wrapper anyway, or at least a sketch of one, and label it as untested!  Or possibly have ARPACK as a different solver option for use in idmrg etc.  It probably isn't efficient because it is too hard to control the number of iterations, but it would be useful in some cases, eg exact diagonalisation.
